service/history/timer_queue_active_task_executor.go

```diff
@@ -310,6 +277,65 @@ Loop:
 	return t.updateWorkflowExecution(ctx, weContext, mutableState, scheduleWorkflowTask)
 }
 
+func (t *timerQueueActiveTaskExecutor) processTimeoutTask(
+	mutableState workflow.MutableState,
+	timerSequenceID workflow.TimerSequenceID,
+	ai *persistencespb.ActivityInfo,
+	updateMutableState *bool,
+	scheduleWorkflowTask *bool,
+) error {
+
+	if timerSequenceID.Attempt < ai.Attempt {
+		//  retry could update activity attempt, should not timeouts new attempt
+		return nil
+	}
+
+	if timerSequenceID.Stamp != ai.Stamp {
```

Comment:
How can this happen? Isn't the timer sequence generated in the same transaction that this function is called?

Reply:
I think you want this in `executeActivityRetryTimerTask` and `transferQueueActiveTaskExecutor.processActivityTask`.

Reply:
Discussed offline, only needed in `executeActivityRetryTimerTask` for now. Pausing the activity is a different story.

Reply:
1. add task.Stamp check earlier.
2.  updated executeActivityRetryTimerTask function - add stamp check.